gr-filter/lib/rational_resampler_impl.cc

```diff
@@ -0,0 +1,304 @@
+        fractional_bw = 0.4;
+    }
+
+    auto d = GR_GCD(interpolation, decimation);
+
+    if (taps.size() && (d > 1)) {
+        GR_LOG_INFO(
+            d_logger,
+            boost::format(
+                "Rational resampler has user-provided taps but interpolation (%1%) and "
+                "decimation (%2%) have a GCD of %3%, which increases the complexity of "
+                "the filterbank. Consider reducing these values by the GCD.") %
+                interpolation % decimation % d);
+    }
+
+    std::vector<TAP_T> staps;
+    if (taps.size() == 0) {
+        if (fractional_bw <= 0) {
+            fractional_bw = 0.4;
+        }
+        // If we don't have user-provided taps, reduce the interp and
+        // decim values by the GCD (if there is one) and then define
+        // the taps from these new values.
+        interpolation /= d;
+        decimation /= d;
+        staps = design_filter(interpolation, decimation, fractional_bw);
+    } else {
+        staps = taps;
+    }
+
+    this->set_relative_rate((uint64_t)interpolation, (uint64_t)decimation);
+    this->set_output_multiple(1);
+
+    d_firs.reserve(interpolation);
+    for (unsigned i = 0; i < interpolation; i++) {
+        d_firs.emplace_back(1, std::vector<TAP_T>());
+    }
+
+    set_taps(staps);
+    install_taps(d_new_taps);
+}
+
+template <class IN_T, class OUT_T, class TAP_T>
+void rational_resampler_impl<IN_T, OUT_T, TAP_T>::set_taps(const std::vector<TAP_T>& taps)
+{
+    d_new_taps = taps;
+    d_updated = true;
+
+    // round up length to a multiple of the interpolation factor
+    int n = taps.size() % this->interpolation();
+    if (n > 0) {
+        n = this->interpolation() - n;
+        while (n-- > 0) {
+            d_new_taps.insert(d_new_taps.end(), 0);
+        }
+    }
+
+    assert(d_new_taps.size() % this->interpolation() == 0);
+}
+
+template <class IN_T, class OUT_T, class TAP_T>
+void rational_resampler_impl<IN_T, OUT_T, TAP_T>::install_taps(
+    const std::vector<TAP_T>& taps)
+{
+    int nfilters = this->interpolation();
+    int nt = taps.size() / nfilters;
+
+    assert(nt * nfilters == (int)taps.size());
+
+    std::vector<std::vector<TAP_T>> xtaps(nfilters);
+
+    for (int n = 0; n < nfilters; n++)
+        xtaps[n].resize(nt);
+
+    for (int i = 0; i < (int)taps.size(); i++)
+        xtaps[i % nfilters][i / nfilters] = taps[i];
+
+    for (int n = 0; n < nfilters; n++)
+        d_firs[n].set_taps(xtaps[n]);
+
+    set_history(nt);
+    d_updated = false;
+}
+
+/**
+ * @brief Given the interpolation rate, decimation rate and a fractional bandwidth,
+    design a set of taps.
+ *
+ * @param interpolation interpolation factor (integer > 0)
+ * @param decimation decimation factor (integer > 0)
+ * @param fractional_bw fractional bandwidth in (0, 0.5)  0.4 works well. (float)
+ */
+template <class IN_T, class OUT_T, class TAP_T>
+std::vector<TAP_T> rational_resampler_impl<IN_T, OUT_T, TAP_T>::design_filter(
```

Comment:
do these design_filter functions even access the rational resampler's state? If not then they should probably either be static functions or (more likely) plain (templated) functions only in the `_impl.cc`?

This may even remove some template complications.